### ISSUE TITLE
Discuss issues with adopting JSON API

### ISSUE THREAD

**Issue by mglaman** [mglaman/commerce_cart_api#11]
Using this to help document and describe our problems with (and why we did not) adopt JSON API for this implementation immediately.

tl;dr tasks identified

- [x] Define the specific ResourceType used in an endpoint, being able to bypass \Drupal\jsonapi\ResourceType\ResourceTypeRepository::get and the entity type + bundle check, and somehow have it always be used
---

Initial branch in which I attempted to use JSON API https://github.com/mglaman/commerce_cart_api/tree/jsonapi

One of the most critical blockers is the fact that JSON API is strict on each API endpoint having a distinct `_entity_type` and `_bundle` definition within the route. When working with content repositories, this makes sense. However, in Drupal Commerce, we do not care about the type of order, order item, or purchasable entity (entity type _and_ bundle.) We have shared interfaces that ensure basic fields and data are available and rely on those for the system to operate.

In `\Drupal\jsonapi\Context\CurrentContext::getResourceType` the `ResourceType` repository will always return  `null` if the bundle is missing. See `\Drupal\jsonapi\ResourceType\ResourceTypeRepository::get`

```
      if ($resource->getEntityTypeId() == $entity_type_id && $resource->getBundle() == $bundle) {
        return $resource;
      }
```

As far as I can tell, there is no way to say "We will use this `ResourceType` for our endpoint".

It looks like this might be partway feasible by ensuring we pass a `ResourceType` to `\Drupal\jsonapi\Normalizer\JsonApiDocumentTopLevelNormalizer::normalize`. But probably not as it is resolved using the ResourceTypeRepository elsewhere (ie: `\Drupal\jsonapi\EntityToJsonApi::calculateContext`)

This logic is what caused the `/carts` path to just die, https://github.com/mglaman/commerce_cart_api/blob/jsonapi/src/Controller/CartResourceController.php#L71. We may have a collection of random cart types.

I know `jsonapi_extras` provides `\Drupal\jsonapi_extras\ResourceType\ConfigurableResourceTypeRepository`. However it is still tied to `%s--%s` as entity type and entity bundle. It also shows that to provide a different ResourceType, the entire ResourceTypeRepository needs to be swapped out (which we cannot do in Commerce.)

It works _nearly perfectly_ except for when normalization occurs and the JSON API context attempts to load the bundle from the route to understand the ResourceType. Have _any other way_ to provide this would solve our needs. Such as allowing a route parameter to define the class/service to be used, 

However, the problem is also that the JSON API seralizers all check `\Drupal\jsonapi\ResourceType\ResourceType::getBundle` for the bundle and not the entity being operated on. 

The main blocker for GET requests would be: `\Drupal\jsonapi\Context\FieldResolver::getFieldItemDefinitions` uses `$bundle = $resource_type->getBundle();`.

Our CRUD operations are more formatted as an RPC (add to cart, patch to change quantities.) So I'm not worried about `denormalize` checking the bundle there. That is fine in the scope of a full admin CRUD.

I have considered using subrequests. Here's what it would look like

**Fetch the collection fo carts the user has: users have a single cart per order type and store**

* Retrieve all of the carts for the user
* Group those by bundle
* Create a subrequest for each bundle route
* Somehow combine the responses
* Send to use

**Fetch the single requested cart**

* Take the cart ID/UUID, validate it belongs to user
* Make a subrequest using the known bundle from loaded order
* Return response.

Honestly the subrequest does not seem horrible. The hard part is somehow conjoining all of requests, which may be near impossib.e

**Comment by mglaman** [comment]
Here is the bug.

```
TypeError: Argument 2 passed to Drupal\jsonapi\Normalizer\JsonApiDocumentTopLevelNormalizer::expandContext() must be an instance of Drupal\jsonapi\ResourceType\ResourceType, null given, called in /Users/mglaman/Drupal/commerce/demo-project/web/modules/contrib/jsonapi/src/Normalizer/JsonApiDocumentTopLevelNormalizer.php on line 211 in Drupal\jsonapi\Normalizer\JsonApiDocumentTopLevelNormalizer->expandContext() (line 254 of /Users/mglaman/Drupal/commerce/demo-project/web/modules/contrib/jsonapi…
```

From

```
commerce_cart_api.cart_collection:
  path: '/jsonapi/cart'
  defaults:
    _controller: '\Drupal\commerce_cart_api\Controller\CartResourceController::getCarts'
  methods: [GET]
  requirements:
    _access: 'TRUE'
    _format: 'json'
    _cart_api: 'TRUE'
  options:
    _auth:
      - cookie
    _is_jsonapi: 1
    # @todo is this needed or autoadded by _is_jsonapi notation
    serialization_class: '\Drupal\jsonapi\Resource\JsonApiDocumentTopLevel'
```

Which does

```
  public function getCarts(Request $request, RouteMatchInterface $routeMatch) {
    $carts = $this->commerceCartCartProvider->getCarts();
    $this->fixInclude($request);

    $entity_collection = new EntityCollection($carts);
    $response = new ResourceResponse(new JsonApiDocumentTopLevel($entity_collection), 200, []);
    return $response;
  }
...
  /**
   * @param \Symfony\Component\HttpFoundation\Request $request
   *
   * @see \Drupal\jsonapi\Normalizer\JsonApiDocumentTopLevelNormalizer::expandContext
   */
  protected function fixInclude(Request $request) {
    // @todo Fix this query param mock.
    // This is read in \Drupal\jsonapi\Normalizer\JsonApiDocumentTopLevelNormalizer::expandContext
    // @todo we also cannot run `include` for included entities (order items)
    $request->query->set('include', 'order_items');

    $request->query->set('fields', [
      'commerce_order--default' => 'uuid,adjustments,total_price,order_items',
      'commerce_order_item--default' => 'uuid,adjustments,unit_price,total_price,quantity,title,purchased_entity',
    ]);
  }
```

**Comment by mglaman** [comment]
I added a custom ResourceType and push it into the request. But this does nothing due to

\Drupal\jsonapi\Context\FieldResolver::resolveInternal

this tries to grab the ResourceType from the ResourceTypeRepository, which gives us no chance to change what is returned.

**Comment by mglaman** [comment]
It _works_ when we do not try to `?include=order_items`.

```
{
  "data": [
    {
      "type": "commerce_order--physical",
      "id": "f4d2360c-c1be-446c-b5b1-717c512e437c",
      "attributes": {
        "order_id": 1,
        "uuid": "f4d2360c-c1be-446c-b5b1-717c512e437c",
        "order_number": null,
        "mail": null,
        "ip_address": "127.0.0.1",
        "adjustments": null,
        "total_price": {
          "number": "38.000000",
          "currency_code": "USD"
        },
        "state": "draft",
        "data": null,
        "locked": false,
        "created": 1523630609,
        "changed": 1523631019,
        "placed": null,
        "completed": null,
        "cart": true,
        "checkout_step": null
      },
      "relationships": {
        "type": {
          "data": {
            "type": "commerce_order_type--commerce_order_type",
            "id": "96be3489-3d5f-46ef-b864-be3a8bb38a8a"
          },
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/type",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/type"
          }
        },
        "store_id": {
          "data": {
            "type": "commerce_store--online",
            "id": "e42f5a27-4ed1-4ebe-8f94-453994077bc0"
          },
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/store_id",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/store_id"
          }
        },
        "uid": {
          "data": {
            "type": "user--user",
            "id": "08198e37-94f1-405e-aad9-b62bd8c59fda"
          },
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/uid",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/uid"
          }
        },
        "billing_profile": {
          "data": null,
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/billing_profile",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/billing_profile"
          }
        },
        "checkout_flow": {
          "data": null,
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/checkout_flow",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/checkout_flow"
          }
        },
        "payment_gateway": {
          "data": null,
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/payment_gateway",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/payment_gateway"
          }
        },
        "payment_method": {
          "data": null,
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/payment_method",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/payment_method"
          }
        },
        "coupons": {
          "data": [],
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/coupons",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/coupons"
          }
        },
        "order_items": {
          "data": [
            {
              "type": "commerce_order_item--physical_product_variation",
              "id": "f170d0f1-deaa-4e26-965c-b53940d04801"
            }
          ],
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/order_items",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/order_items"
          }
        },
        "shipments": {
          "data": [],
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/shipments",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/shipments"
          }
        }
      },
      "links": {
        "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c"
      }
    }
  ],
  "jsonapi": {
    "version": "1.0",
    "meta": {
      "links": {
        "self": "http:\/\/jsonapi.org\/format\/1.0\/"
      }
    }
  },
  "links": {
    "self": "http:\/\/localhost:32770\/jsonapi\/cart?XDEBUG_SESSION_START=1"
  }
}

```

**Comment by mglaman** [comment]
Once I fixed the access

```
{
  "data": [
    {
      "type": "commerce_order--physical",
      "id": "f4d2360c-c1be-446c-b5b1-717c512e437c",
      "attributes": {
        "order_id": 1,
        "uuid": "f4d2360c-c1be-446c-b5b1-717c512e437c",
        "order_number": null,
        "total_price": {
          "number": "38.000000",
          "currency_code": "USD"
        }
      },
      "relationships": {
        "store_id": {
          "data": {
            "type": "commerce_store--online",
            "id": "e42f5a27-4ed1-4ebe-8f94-453994077bc0"
          },
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/store_id",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/store_id"
          }
        },
        "order_items": {
          "data": [
            {
              "type": "commerce_order_item--physical_product_variation",
              "id": "f170d0f1-deaa-4e26-965c-b53940d04801"
            }
          ],
          "links": {
            "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/relationships\/order_items",
            "related": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c\/order_items"
          }
        }
      },
      "links": {
        "self": "http:\/\/localhost:32770\/jsonapi\/commerce_order\/physical\/f4d2360c-c1be-446c-b5b1-717c512e437c"
      }
    }
  ],
  "jsonapi": {
    "version": "1.0",
    "meta": {
      "links": {
        "self": "http:\/\/jsonapi.org\/format\/1.0\/"
      }
    }
  },
  "links": {
    "self": "http:\/\/localhost:32770\/jsonapi\/cart?XDEBUG_SESSION_START=1"
  }
}

```

**Comment by mglaman** [comment]
The next blocker is \Drupal\jsonapi\Normalizer\JsonApiDocumentTopLevelNormalizer::expandContext

in

```
    $public_includes = array_map(function ($include_str) use ($resource_type) {
      $resolved = $this->fieldResolver->resolveInternal(
        $resource_type->getEntityTypeId(),
        $resource_type->getBundle(),
        trim($include_str)
      );
      // We don't need the entity information for the includes. Clean it.
      return preg_replace('/\.entity\./', '.', $resolved);
    }, $includes);
```

The resource type has a bundle which is null. Made a patch to get JSON API to fix.

**Comment by mglaman** [comment]
PATCH https://www.drupal.org/project/jsonapi/issues/2960766

**Comment by mglaman** [comment]
From Slack

> gabesullice [10:30 AM]
> @mglaman I did see that, I need to take a look at the patch and see what's happening there. FWIW, it'll still be outside the realm of support. But it's interesting to see a relatively minor change is the only thing that blocks heterogenous collections
> 
> I wonder if the better patch might be to pass in an array of ResourceTypes instead of NULL ->bundle()

**Comment by mglaman** [comment]
One consideration from gabesuillice

> I really think you should follow the spec but maybe not use the module

But that is near impossible, as we would be recreating the JSON API module, then.

**Comment by wimleers** [comment]
Good news: https://www.drupal.org/project/jsonapi/issues/2960766#comment-12685668
